test/cql-pytest/nodetool.py

```diff
@@ -0,0 +1,68 @@
+# Copyright 2021 ScyllaDB
+#
+# This file is part of Scylla.
+#
+# Scylla is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Scylla is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+##################################################################
+
+# This file provides a few nodetool-like commands that may be useful
+# for tests. The intent is *not* to test nodetool itself, but just to
+# use its functionality to test CQL. So when testing against a locally-
+# running Scylla these functions do not actually use nodetool but rather
+# the REST API. This simplifies running the test because an external Java
+# process implementing JMX is not needed. But when the REST API port is
+# not available (running against Cassandra or some remote installation of
+# Scylla) "nodetool" *is* used, and must be available in the path or
+# chosen with the NODETOOL environment variable.
+
+import requests
+import os
+import subprocess
+
+# For a "cql" object connected to a particular node, find the REST API URL
+# with the same node and port 10000.
+def rest_api_url(cql):
+    return f'http://{cql.cluster.contact_points[0]}:10000'
+
+# Check whether the REST API at port 10000 is available - if not we will
+# fall back to using an external "nodetool" program.
+# We only check this once per REST API URL, and cache the decision.
+checked_rest_api = {}
+def has_rest_api(cql):
+    url = rest_api_url(cql)
+    if not url in checked_rest_api:
+        # Scylla's REST API does not have an official "ping" command,
+        # so we just list the keyspaces as a (usually) short operation
+        try:
+            ok = requests.get(f'{url}/column_family/name/keyspace').ok
+        except:
+            ok = False
+        checked_rest_api[url] = ok
+    return checked_rest_api[url]
+
+
+# Run the external nodetool executable (can be overridden by the NODETOOL
+# environment variable). Only call this if the REST API doesn't work.
+nodetool_cmd = os.getenv('NODETOOL') or 'nodetool'
+def run_nodetool(cql, *args):
+    host = cql.cluster.contact_points[0]
+    subprocess.run([nodetool_cmd, '-h', host, *args])
+
+def flush(cql, table):
+    ks, cf = table.split('.')
+    if has_rest_api(cql):
```

Comment:
As we add more commands it would be nice to hide this boilerplate in a function and just call:
```python
do_invoke_nodetool(url="...", command=[...])
```
Can be deferred to when we actually add more commands.

Reply:
But what boilderplate? There is just an 'if' with two cases here, I think it's clearer than a function with two parameters. Also please note that the "url" part is slightly more complicated than what you suggested - there is also the method ("post"), and the parameters (which aren't part of the URL) - which is all written as a `requests.post(...)` command below, which isn't really boilerplate.

Reply:
Even a single statement can be boiler plate if it has to copied to all commands. But I agree I'm stretching it a bit here.